## ANLY-580 Week 3 Homework
## Yi Li

#### 9. Save some text into a file corpus.txt. Define a function load(f) that reads from the file named in its sole argument, and returns a string containing the text of the file.

* Use nltk.regexp_tokenize() to create a tokenizer that tokenizes the various kinds of punctuation in this text. Use one multi-line regular expression, with inline comments, using the verbose flag (?x).

* Use nltk.regexp_tokenize() to create a tokenizer that tokenizes the following kinds of expression: monetary amounts; dates; names of people and organizations. (Only need to use different characters E.g. capitalization)

In [1]:
# Define a function load(f) that reads from the file named in its sole argument, 
# and returns a string containing the text of the file.
def load(f):
    file = open(f)
    return(file.read())

load('corpus.txt')[:1000]

'Student\'s t-test\nFrom Wikipedia, the free encyclopedia\nThe t-test is any statistical hypothesis test in which the test statistic follows a Student\'s t-distribution under the null hypothesis.\nA t-test is most commonly applied when the test statistic would follow a normal distribution if the value of a scaling term in the test statistic were known. When the scaling term is unknown and is replaced by an estimate based on the data, the test statistics (under certain conditions) follow a Student\'s t distribution. The t-test can be used, for example, to determine if two sets of data are significantly different from each other.\n\nWilliam Sealy Gosset, who developed the "t-statistic" and published it under the pseudonym of "Student".\nThe t-statistic was introduced in 1908 by William Sealy Gosset, a chemist working for the Guinness brewery in Dublin, Ireland. "Student" was his pen name.[1][2][3][4]\nGosset had been hired owing to Claude Guinness\'s policy of recruiting the best graduat

In [2]:
# create a tokenizer that tokenizes the various kinds of punctuation in this text 
import nltk
text = load('corpus.txt')
pattern = r'''(?x)    # set flag to allow verbose regexps
    [^\w\d\s]  # Any non-alphanumeric, non-digit, non-whitespace character
'''
set(nltk.regexp_tokenize(text, pattern))

{'"',
 '#',
 '%',
 '&',
 "'",
 '(',
 ')',
 '+',
 ',',
 '-',
 '.',
 '/',
 ':',
 ';',
 '=',
 '?',
 '[',
 '\\',
 ']',
 '^',
 '{',
 '}',
 '®',
 '–',
 '…',
 '−'}

In [3]:
# create a tokenizer that tokenizes monetary amounts
text = 'Costs: $12.40, $1, $500, $9.99...'
pattern = r'\$?\d+\.?\d*%?'
nltk.regexp_tokenize(text, pattern)

['$12.40', '$1', '$500', '$9.99']

In [4]:
# create a tokenizer that tokenizes dates 
text = '''This page was last edited on 22 September 2017, at 12:58.
    Other sample dates: 09/27/2017, 09-27-2017, 09272017 ...
'''
pattern = r'''(?x)
    \d{2}[\-|\/]?\d{2}[\-|\/]?\d{2,4}   # match "09/27/2017", "09-27-2017", "09272017"
    |\d{1,2}\s[A-Z]\w+[\.|\,|\s]+\d+    # match "22 September 2017"
'''
nltk.regexp_tokenize(text, pattern)

['22 September 2017', '09/27/2017', '09-27-2017', '09272017']

In [6]:
# creater a tokenizer that tokenizes names of people and organizations
# (find capitalized words in sequence)
text = '''William Sealy Gosset and William A. Silverman: Two "students" of science". Pediatrics. 
Wikipedia® is a registered trademark of the Wikimedia Foundation, Inc., a non-profit organization.
'''
pattern = r'''(?x) # set flag to allow verbose regexps
  (?:[A-Z]\w*\S?\s?)+ # find capitalized words in sequence
'''
nltk.regexp_tokenize(text, pattern)

['William Sealy Gosset ',
 'William A. Silverman: Two ',
 'Pediatrics. ',
 'Wikipedia® ',
 'Wikimedia Foundation, Inc.']

#### 43. With the help of a multilingual corpus such as the Universal Declaration of Human Rights Corpus (nltk.corpus.udhr), and NLTK's frequency distribution and rank correlation functionality (nltk.FreqDist, nltk.spearman_correlation), develop a system that guesses the language of a previously unseen text. For simplicity, work with a single character encoding and just a few languages.

In [7]:
import nltk
from nltk.corpus import udhr

# create a function that return a list of the rank of each letter in a certain text
def RankList(text):
    fd = nltk.FreqDist(text.lower())
    dist = sorted([(count, char) for char, count in dict(fd).items() if char.isalpha()], reverse = True)
    rank = [(char, i+1) for i, (_, char) in enumerate(dist)]
    return rank

# use the following corpus to generate the rank of letters of different languages
eng = udhr.raw('English-Latin1')
ger = udhr.raw('German_Deutsch-Latin1')
chi = udhr.raw('Chickasaw-Latin1')

sample = load('corpus.txt')

#print(RankList(eng), RankList(ger), RankList(chi))
print('English:', nltk.spearman_correlation(RankList(eng), RankList(sample)))
print('German:', nltk.spearman_correlation(RankList(ger), RankList(sample)))
print('Chickasaw:', nltk.spearman_correlation(RankList(chi), RankList(sample)))

English: 0.9494017094017094
German: 0.7065384615384616
Chickasaw: 0.04573687182382835


So the sample's language is English since the probability is the highest with the value 94.94%.